### ISSUE TITLE
Diagnostic tooling

### ISSUE THREAD

**Issue by cadorn** [ragents/ragents#1]
How do you envision the diagnostic tooling working?

I am interested in helping contribute to a dev setup that allows for easy testing of apps built using ragents and the messages passed between them. I have some tooling I can integrate and you could run to try things out (can for example deploy test app to many nodes and wire up a central logging channel for analysis using minimal code). I am doing something similar with other developers including the firebug working group with the goal to validate my open source tooling further. To start with I would be integrating [Bash.Origin](https://github.com/bash-origin/bash.origin) and then configure a few scripts in `bin` to load various open source tooling to start runtime containers and load the source code etc... The plugins are all foundational pieces that the more abstracted devtools will drive later and the whole thing can be seeded today using `bash.origin` scripts.

I am interested in ragents because it fits into the communication model of my system where I am looking at treating each piece of the architecture as a node. I am interested in what your approach has to offer and the patterns it enables for distributed apps. I am looking at integrating the tooling to bring these aspects to light.

I welcome feedback on `bash.origin` and aim to make it a vehicle to enable these kinds of experiments in other projects without needing the original project to make any changes to its source code.

**Comment by pmuellr** [comment]
Hi Chris!

Right now, my plan is to just do some ad hoc tooling.  Hadn't really been thinking big picture for that story yet.

For instance, ragents is very similar to the communication framework I built for [weinre](http://people.apache.org/~pmuellr/weinre-docs/latest/).  That was mobile, but now I do more cloud development, and ... running into some of the same problems - difficult to be running ANOTHER server (to handle diagnostic bits) beside my already running server.  So ... wanted to split the transport out, to see if it would be useful there as well.

In terms of the tooling I was thinking about, figured I'd do some health monitors (memory usage), see if I can do some kind of a REPL integration, and then look at doing v8 debug integration so get a better story for "cloud debugging" than the current state of the art (eg, how can I debug an app that is running > 1 instance behind a proxy or whatever).

**Comment by cadorn** [comment]
>  ragents is very similar to the communication framework I built for weinre

Are you looking at using ragents within weinre eventually? Excuse my ignorance. Its been a while since I looked at weinre .

> difficult to be running ANOTHER server (to handle diagnostic bits) beside my already running server

Interesting. Can you expand on this a bit more.

My research has been going into how to structure things in-process so it can be automatically split out into multiple processes when scaling/distributing. This would be the focus of my tooling integration. To support running ragents in all sorts of different setups to see how the concepts adjusts and scales.

> In terms of the tooling I was thinking about

So you are thinking about implementing tooling with ragents instead of for ragents?

If with ragents then I am working on similar tooling in my own work and with the firebug working group. We are working on generic debug widgets that can be fed from servers. The widgets are connected across nodes using ids etc... This would fit into your model as well.

Maybe we can have a quick call to see what your plans are and where we may overlap. If you can optionally re-use some of our components we can validate your approach as well as ours.

**Comment by pmuellr** [comment]
> Are you looking at using ragents within weinre eventually? Excuse my ignorance.
> Its been a while since I looked at weinre .

The theory is, I **could** rip out the equivalent plumbing in weinre and replace it with ragents.  But there's no value in doing that, so .. I'm not going to do it.  At some point, I may make some v8 debug agent which could be used to talk to node-inspector (or similiar).  This would be useful for Cloud Foundry apps, as CF apps only get one port open for their app, which is presumably running their primary web server - so can't open **another port** (exposed publicly) as a v8 debug port.

> So you are thinking about implementing tooling with ragents instead of for ragents?

Yes, **with** ragents, not **for** ragents.  Hopefully ragents is simple enough, it won't need tooling!

> If with ragents then I am working on similar tooling in my own work and with the firebug working group. 

Totally up for a call.  Do you have a link to any description or code for this work?

**Comment by cadorn** [comment]
> I may make some v8 debug agent ... apps only get one port open for their app

Interesting. This is a common issue. Works assuming you can load your own binary extensions/compile nodejs on host if I understand correctly. So it would be a way to connect to v8 debug through the HTTP/websocket port using a special header byte sequence or something?

> Do you have a link to any description or code for this work?

I am on the cusp of making various projects of mine more accessible. Until then it may be difficult to form a big-picture without a conversation. Here are some links:
- **Widgetization** http://fireconsole.github.io/harviewer/ (click on `Embed`) as a proof of concept wrap the AMD harviewer project using metadata and static analysis and make it accessible using a non-conflicting [code loader](https://github.com/pinf/pinf-loader-js). Has worked with the ACE code editor [as well in the past](https://github.com/cadorn/ace-extjs). Need to get working again. This works for all AMD projects augmented with sufficient declarations to fill gaps in `package.json` of the respective packages.
- **Systemization** https://github.com/firebug/firebug.io based on http://devcomp.io which is a clonable dev system which is about to get a simplification refactor as I bring in more clean primitives. I am working on half a dozen clones of devcomp and fine-tuning all the various abstractions needed for devtools to manipulate a system.

The point is to be able to link in devtools into your project and structure them using the `package.json` files of the various packages in your project. The metadata must sufficiently describe the API or internals of the code within the respective package to allow static analysis tools and runtime injection monitoring calls to fill in the rest. Many times, declaring entry points, what is consumed and exported and some internal properties to watch is sufficient.

My objective with ragents would be to put in place an environment that would allow anyone by declaring specific properties in their package.json files, load ragents based dev widgets into any supporting IDE/debugger like firebug.next which I have intimate access to and am working on the [next version for FirePHP](https://github.com/firephp/firephp-for-firebug.next) using the above technology as well. Even if the widgets get loaded in a devtools client and wired up without the UI, the UI of the widgets is always optionally accessible to potentially provide a more complete picture of the component that a standardized devtool might provide. i.e. The widget approach is a super-set of the api approach and enhances the api approach significantly.

There are a plethora of permutations possible once the above is achieved and my goal is to fine-tune the integration and learning process of the open source tooling I have been working on in an effort to make it more accessible.

You can reach me on skype christoph@christophdorn.com or we can connect elsewhere.

**Comment by pmuellr** [comment]
> Interesting. This is a common issue. Works assuming you can load your own binary extensions/compile nodejs on host if I understand correctly. So it would be a way to connect to v8 debug through the  HTTP/websocket port using a special header byte sequence or something?

The basic thought, for v8 debug, is that I'd arrange to launch a second node process, which would read the v8 debug port, and then create a "agent" for that debug session.  The agent would just forward messages to/from the v8 debug port, out through the agent.  The messaging that ragents supports actually EXACTLY matches what the v8 debug protocol does (request/responses, and events), basically same as chrome debug protocol.  Again, same basic idea as weinre, only ... different.

Having said that, you should probably think of ragents as something like [DBus](https://pythonhosted.org/txdbus/dbus_overview.html), only it sends JSON messages over WebSocket, so it's easy to use in a browser, node.js, and any other place you can get access to a WebSocket.

RAgents, or something similiar, might work well for your needs, as it's fairly simple to build agents, send messages to them / get events from them, all as (in the end) WebSocket clients, using a generic server to route the messages between the goodies.

If you're just dealing with things locally - there's no cross-machine communication going on - then RAgents is overkill, because you could use local sockets instead.  Or DBus :-)  The overkill isn't so bad, message throughput is actually quite good when running intra-machine, you just don't really need WebSocket for that.

**Comment by cadorn** [comment]
> RAgents, or something similiar, might work well for your needs, as it's fairly simple to build agents, send messages to them / get events from them, all as (in the end) WebSocket clients, using a generic server to route the messages between the goodies.

Right. The idea is to have widgets with backend and front end-piece that are automatically wired up when they boot. The containers within which they boot are pre-configured to allow connections between certain endpoints on different nodes and can act as the dumb websocket relays. I am setting up similar communication channels with @kriskowal 's `qi-io` ecosystem as well as others. My objective is to arrive at a widget interface that allows connecting to all sorts of endpoints natively and as low-level as we can get it in pure JS using a common JSON structure to describe the connections.

We can then write dev widgets that seek the best connection based on their needs at any point in time. e.g. Initiate a new direct websocket connection to a different IP using a port that was just provisioned for this temporary purpose to bypass a bunch of proxies when doing a high-log-volume debug session on a secure remote server.
